text/3305-static-in-pattern.md

```diff
@@ -0,0 +1,285 @@
+    }
+}
+```
+
+Needless to say, this is unlikely to ever be upstreamed. Allowing statics in patterns would solve this use-case much more cleanly.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Rust patterns can refer to constants.
+
+```rust
+const EVERYTHING: i32 = 42;
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        EVERYTHING => println!("have all of it"),
+        _ => println!("need moar"),
+    }
+}
+```
+
+With this feature, they can refer to statics as well.
+
+```rust
+static EVERYTHING: i32 = 42;
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        EVERYTHING => println!("have all of it"),
+        _ => println!("need moar"),
+    }
+}
+```
+
+Mutable statics are not allowed, however. Patterns can't reference information that can change at runtime, and also can't be `unsafe`.
+
+```rust
+
+static mut EVERYTHING: i32 = 42;
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        // ERROR can't refer to mutable statics in patterns
+        /* EVERYTHING => println!("have all of it"), */
+        _ => println!("need moar"),
+    }
+}
+```
+
+Statics from `extern` blocks are allowed, but they must be marked as trusted using the (not-yet-implemented) [trusted external statics](https://github.com/rust-lang/lang-team/issues/149) feature.
+
+```rust
+extern "C" {
+    #[unsafe(trusted_extern)]
+    static EVERYTHING: i32;
+}
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        EVERYTHING => println!("have all of it"),
+        _ => println!("need moar"),
+    }
+}
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+For a static to be eligible for use in a pattern, it must:
+
+- not be marked `mut`
+- not be marked `#[thread_local]`
+- not come from an extern block, unless it is marked as safe to use with the [trusted external statics](https://github.com/rust-lang/lang-team/issues/149) feature
+- have a type that satisfies the structural match rules, as described in [RFC 1445](1445-restrict-constants-in-patterns.md), but without any allowances for backward compatibility like there are for consts (e.g., floating point numbers in patterns) . These rules exclude all statics with interior mutability.
+
+Static patterns match exactly when a const pattern with a const of the same type and value would match.
+
+The values of statics are treated as opaque for reachability and exhaustiveness analysis.
+
+```rust
+static TRUE: bool = true;
+static FALSE: bool = false;
+
+fn foo(scrutinee: bool) {
+    match scrutinee {
+        TRUE | FALSE => println!("bar"),
+
+        // The compiler will throw an error if you remove this branch;
+        // it is not allowed to look into the values of the statics
+        // to determine that it is unreachable.
+        _ => println!("baz"),
+    }
+}
+```
+
+As an exception, when all safe values of a type are structurally equal, the compiler is allowed to see that the match will always succeed.
```

Comment:
"all safe values" is definitely impossible to implement. I could define
```rust
// Safety invariant: the field is always 0.
struct MyWeirdUnit(i32);
```
`MyWeirdUnit` has only one safe value, but the compiler has no way of knowing that.

Maybe you meant "all valid values", but even that seems way too subtle to include here. Do we have any precedence for a rule like that?

Reply:
Also, "is allowed to see" is too weak; I assume you mean that the compiler *must* accept the code below, not just that it is allowed to accept the code below.

Reply:
And finally, given that this is a large fraction of the RFC, there also needs to be motivation for all this complexity. The motivating example certainly doesn't need any of this.

Reply:
I basically interpreted this as exclusively applying to ZSTs, but wanting to be more general in case we somehow added the ability to increase the number of cases we could detect this.

Reply:
For example, is isn't possible now, but it eventually could be possible to mark fields as ignored when deriving `Eq`, which would allow such a type to be non-zero-sized while still having a zero-sized equivalence class.

Reply:
I guess the concept I am trying to express is "all valid values constructible in safe code, ignoring visibility of types and fields." So @RalfJung's `MyWeirdUnit` struct has 2^32 "safe" values for the purposes of this analysis. But a `&&()` that is actually null is not a "safe" value (this is the case I was trying to account for).

Reply:
Yes, this should be "valid". I had a reason in my head for why this should be "safe", but upon writing it out I realized that my logic made no sense.

As stated in the RFC, there's no practical use-case (AFAIK) for the "one valid value" rule. It's just that, on general principle, I feel the compiler should never force you to write useless code that it statically knows to be unreachable, unless there's a good justification for it.